File: Daily_Lists/search_in_a_Big_sorted_array/search_in_a_big_sorted_array.cpp
/**
 * Definition of ArrayReader:
 *
 * class ArrayReader {
 * public:
 *     int get(int index) {
 *          // return the number on given index,
 *          // return -1 if index is less than zero.
 *     }
 * };
 */
class Solution {
public:
    /**
     * @param reader: An instance of ArrayReader.
     * @param target: An integer
     * @return: An integer which is the first index of target.
     * @author: Egbert Li
     * @link: http://www.lintcode.com/en/problem/search-in-a-big-sorted-array/
     */
    int searchBigSortedArray(ArrayReader *reader, int target) {
        int index = 1;
        while (target > reader->get(index - 1)) {
            index *= 2;
        }
        int start = 0;
        int end = index - 1;
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (target > reader->get(mid)) {
                start = mid;
            } else {
                end = mid;
            }
        }
        if (target == reader->get(start)) {
            return start;
        } else if (target == reader->get(end)) {
            return end;
        } else {
            return -1;
        }
    }
};

File: Yixuan/16_01_17/search_in_a_big_sorted_array.py
"""
Definition of ArrayReader:
class ArrayReader:
    def get(self, index):
        # this would return the number on the given index
        # return -1 if index is less than zero.
"""
class Solution:
    # @param {ArrayReader} reader: An instance of ArrayReader
    # @param {int} target an integer
    # @return {int} an integer
    # @author: Egbert Li
    # @link: http://www.lintcode.com/en/problem/search-in-a-big-sorted-array/
    def searchBigSortedArray(self, reader, target):
        index = 1
        while target > reader.get(index - 1):
            index *= 2
        start = 0
        end = index - 1
        while start + 1 < end:
            mid = start + int((end - start) / 2)
            if target > reader.get(mid):
                start = mid
            else:
                end = mid
        if target == reader.get(start):
            return start
        elif target == reader.get(end):
            return end
        else:
             return -1
